### PR TITLE
[Merged by Bors] - feat: Infinite products

### PR DESCRIPTION
**Pull request by trivial1711**
Last year, YaelDillies made a pull request to mathlib3 that unfortunately never got merged in: https://github.com/leanprover-community/mathlib/pull/18405. This is the mathlib4 version of that pull request.

We define arbitrarily indexed products in a commutative monoid with a topology. This is done by "multiplicativizing" the currently existing definitions and theorems about arbitrarily indexed sums. That is, the existing code is rewritten in the multiplicative setting, and the original definitions and theorems are recovered using `@[to_additive]`. Please see this thread on Zulip for information on why this approach was chosen: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Infinite.20products

As YaelDillies wrote in the description of https://github.com/leanprover-community/mathlib/pull/18405, there are a few small technical issues that arise when directly "multiplicativizing" theorems in this way:
- I have renamed `cauchySeq_finset_iff_vanishing` to `cauchySeq_finset_iff_sum_vanishing` to make the name multiplicativizable. This is slightly different from the name `cauchy_seq_finset_sum_iff_vanishing` that YaelDillies used, and it is meant to parallel the existing name `cauchySeq_finset_iff_tsum_vanishing`.
- Currently, on master, there is a theorem called `tsum_sum` about taking the `tsum` of a `sum`, and a theorem called `tsum_prod` about taking a `tsum` on a product of two index sets. I have called the multiplicative versions `tprod_of_prod` and `tprod_prod`. This is slightly different from the names `tprod_prod''` and `tprod_prod` that YaelDillies used. eric-wieser suggested renaming `tsum_prod` to `tsum_prod_index` to get around this issue, which I can do in a separate pull request.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (YaelDillies) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:315]: Why is there no multiplicative docstring here?

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (YaelDillies) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:591]: Similarly here

File: Mathlib/Topology/Algebra/InfiniteSum/Defs.lean
Comment (YaelDillies) [inline Mathlib/Topology/Algebra/InfiniteSum/Defs.lean:23]:
```suggestion
## Implementation notes

This is probably not the definition you expect for multiplication because blablabla.
```

File: Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean
Comment (YaelDillies) [inline Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean:92]:
```suggestion
theorem comp_nat_add {f : ℕ → M} {k : ℕ} (h : Multipliable fun n ↦ f (n + k)) : Multipliable f :=
```

File: Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean
Comment (YaelDillies) [inline Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean:113]:
```suggestion
taking a supremum. -/
```

File: Mathlib/Topology/Algebra/InfiniteSum/Order.lean
Comment (YaelDillies) [inline Mathlib/Topology/Algebra/InfiniteSum/Order.lean:151]:
```suggestion
@[to_additive tsum_nonneg]
theorem one_le_tprod (h : ∀ i, 1 ≤ g i) : 1 ≤ ∏' i, g i := by
```

File: Mathlib/Topology/Algebra/InfiniteSum/Order.lean
Comment (YaelDillies) [inline Mathlib/Topology/Algebra/InfiniteSum/Order.lean:209]:
```suggestion
@[to_additive tsum_pos]
theorem one_lt_tprod (hsum : Multipliable g) (hg : ∀ i, 1 ≤ g i) (i : ι) (hi : 1 < g i) :
```

**Review by YaelDillies — COMMENTED** [review]
Please avoid mentioning people *above* the `---` in your PR description as then the mentions will happen every time your commit appears in a merge commit anywhere on github.

Mostly looks good, but I need a second pair of eyes because the diff is big.

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (trivial1711) [reply Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:315]: The reason is silly. It is because I multiplicativized this file before multiplicativizing on `Topology.Algebra.InfiniteSum.Group`, so the lemma `HasProd.update`, which I needed to reference in the multiplicative docstring, didn't exist yet. Thank you for catching this.

**Comment by trivial1711** [comment]
Thank you for the review. I have taken into account your comments.

**Comment by trivial1711** [comment]
Thanks; I've added the docstrings

File: Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean
Comment (YaelDillies) [inline Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean:525]: Missing docstring

File: Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean
Comment (YaelDillies) [inline Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean:532]: Missing docstring

**Review by YaelDillies — COMMENTED** [review]
Okay I think those are the last two

**Comment by trivial1711** [comment]
Sorry for missing so many of them.

**Review by YaelDillies — APPROVED** [review]
Looks alright to me but I need a second pair of eyes because the diff is big.

maintainer merge

Btw no need to ask for Johannes' review. He has left the community in 2019.

File: Mathlib/Topology/Algebra/InfiniteSum/Defs.lean
Comment (sgouezel) [inline Mathlib/Topology/Algebra/InfiniteSum/Defs.lean:64]: Could you explain in the docstring what this actually is, in addition to referring to `hasSum`? In a tooltip, the user will not have the opportunity to see the definition of `hasSum`, so docstrings should be self-contained when possible.

File: Mathlib/Topology/Algebra/InfiniteSum/Defs.lean
Comment (sgouezel) [inline Mathlib/Topology/Algebra/InfiniteSum/Defs.lean:97]: adjust to the multiplicative world.

File: Mathlib/Topology/Algebra/InfiniteSum/Group.lean
Comment (sgouezel) [inline Mathlib/Topology/Algebra/InfiniteSum/Group.lean:359]:
```suggestion
/-- Product divergence test: if `f` is unconditionally multipliable, then `f x` tends to one along
```

File: Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean
Comment (sgouezel) [inline Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean:80]:
```suggestion
theorem hasProd_iff_tendsto_nat [T2Space M] {f : ℕ → M} (hf : Multipliable f) :
```

File: Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean
Comment (sgouezel) [inline Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean:243]:
```suggestion
multipliability assumption on `f`, as otherwise all such products are one. -/
```

**Review by sgouezel — COMMENTED** [review]
bors d+
Thanks so much for doing this!

**Comment by sgouezel** [comment]
(Before merging, please take into account Yael's comment on names in the PR description)

**Comment by YaelDillies** [comment]
(This was already done)

**Pull request closed without merging**
